src/io.jl

```diff
@@ -11,68 +39,101 @@ function readimage(filename::String, transforms::Int = 0)
 
     width = ccall((:png_get_image_width, "libpng"), UInt32, (Ptr{Void}, Ptr{Void}), png_ptr, info_ptr)
     height = ccall((:png_get_image_height, "libpng"), UInt32, (Ptr{Void}, Ptr{Void}), png_ptr, info_ptr)
+    color_type = ccall((:png_get_color_type, "libpng"), UInt32, (Ptr{Void}, Ptr{Void}), png_ptr, info_ptr)
+    bit_depth = ccall((:png_get_bit_depth, "libpng"), UInt32, (Ptr{Void}, Ptr{Void}), png_ptr, info_ptr)
     num_channels = ccall((:png_get_channels, "libpng"), UInt8, (Ptr{Void}, Ptr{Void}), png_ptr, info_ptr)
 
-    rows = ccall((:png_get_rows, "libpng"), Ptr{Ptr{UInt8}}, (Ptr{Void}, Ptr{Void}), png_ptr, info_ptr)
-
-    image = zeros(UInt8, height, width, num_channels)
-    for i = 1:height
-        row = unsafe_load(rows, i)
-        for j = 1:width
-            for c = 1:num_channels
-                image[i, j, c] = unsafe_load(row, num_channels * (j - 1) + c)
-            end
-        end
+    even_depth = ((bit_depth + 1) >> 1) << 1
+    if bit_depth <= 8
+        T = Normed{UInt8, 8}
+    elseif bit_depth <= 16
+        T = Normed{UInt16, even_depth}
+    else
+        # TODO UInt32?
+        error("Unknown bit_depth: $bit_depth")
     end
 
+    colors_type = map_color(color_type, T)
+
+    buf = Array{colors_type}(height, width)
+
+    get_image_pixels!(rawview(channelview(buf)), png_ptr, info_ptr)
+
     png_destroy_read_struct(png_ptr, info_ptr)
 
     close_png(fp)
 
-    return image
+    return buf
 end
 
-# Colors RGB
-function write_rows{T<:FixedPoint}(image::AbstractArray{RGB{T}}, png_ptr::Ptr{Void}, info_ptr::Ptr{Void})
-    channel_view = reinterpret(UInt8, image)
-    for row = 1:size(image, 1)
-        row_buf = channel_view[:, row, :]
-        ccall((:png_write_row, :libpng), Void, (Ptr{Void}, Ptr{UInt8}), png_ptr, row_buf)
+function get_image_pixels!{T<:Unsigned}(buf::AbstractArray{T, 2}, png_ptr::Ptr{Void}, info_ptr::Ptr{Void})
+    height, width = size(buf)
+    rows = ccall((:png_get_rows, "libpng"), Ptr{Ptr{T}}, (Ptr{Void}, Ptr{Void}), png_ptr, info_ptr)
```

Comment:
This and various other ccall locations potentially should use `libpng` the variable (from deps.jl) rather than the hardcoded `"libpng"` or `:libpng`.  That way the name detected by BinDeps will be used.

Reply:
Great thanks!